onnx/onnx.in.proto

```diff
@@ -196,6 +204,112 @@ message NodeProto {
   optional string doc_string = 6;
 }
 
+// Training information
+// TrainingInfoProto stores information for training a model.
+// In particular, this defines two functionalities: an initialization-step
+// and a training-algorithm-step. The semantics of the initialization-step is
+// that all initializers (in ModelProto.graph and in all
+// TrainingInfoProto.algorithm) are first initialized as specified by the
+// initializers in the graph, and then by the "initialization_binding" in every
+// instance in ModelProto.training_info. The semantics of the
+// training-algorithm-step is that all the initializers are updated as
+// specified by the "update_binding" in every instance in training_info.
+//
+// The field "algorithm" defines a computation graph similar to the inference
+// one (i.e., ModelProto.graph) but represents a training algorithm's step.
+// Each execution of a TrainingInfoProto.algorithm leads to the update of some
+// initializers in its TrainingInfoProto.algorithm.initializer and
+// ModelProto.graph.initializer.
+//
+// If there are multiple TrainingInfoProto's, sequentially executing all of
+// them once constitutes a full training algorithm. The initializer update
+// happens at the end of each TrainingInfoProto's execution. If the targeted
+// training algorithm contains consecutive update stages (such as block
+// coordinate descent methods), the user needs to create a TrainingInfoProto
+// for each stage.
+//
+// In addition to training (improving the model based on input data),
+// TrainingInfoProto can store an initialization process (resetting the model
+// back to its original state as if no training has been consumed) as a
+// computation graph in the "initialization" field. The execution of
+// "initialization" is similar to that of "algorithm". The user can use
+// "initialization_binding" to assign the outputs of "initialization" to
+// ModelProto.graph.initializer and TrainingInfoProto.algorithm.initializer.
+// 
+// Sveral inference components are usable in TrainingInfoProto; the list
+// includes
+//   1. ModelProto.graph is a callble inside TrainingInfoProto.algorithm.
+//      Each call is a NodeProto in TrainingInfoProto.algorithm.node with
+//      NodeProto.op_type == "Graph" and
```

Comment:
Should we document `Graph` as a new operator?

Reply:
Why not just call a graph using its name as the "op_type"?

Reply:
@gramalingam, I would like to make the Graph callable even if it's name is `LSTM`?

Reply:
I agree on using the graph name as the “op_type", which is much more extendable and also following the same naming convention as "Function".

Reply:
@TMVector, It seems we need to use `op_type=ModelProto.graph.name`. Because the names of `NodeProto`s must be unique in a single graph, if we do `op_type=Graph + op_name=ModelProto.graph.name`, we can only call the inference once in a training graph.

Let's consider calling `InferenceGraph` twice as an example. If we use `op_type=Graph + op_name=ModelProto.graph.name`, we will have
```
X --> Node(name="MyInferenceGraphName", op_type="Graph") -->Y
X_1 --> Node(name="MyInferenceGraphName", op_type="Graph") -->Y_1
```
which breaks the uniqueness of operator names. Maybe it's not a problem but we need another discussion to see if everyone agrees on it.

In contrast, if we do `op_type=ModelProto.graph.name`, we can have
```
X --> Node(name="CallMyGraph", op_type="MyInferenceGraphName") -->Y
X_1 --> Node(name="CallMyGraphAgain", op_type="MyInferenceGraphName") -->Y_1
```

Reply:
We are in complete agreement that we shouldn't use `NodeProto.name` to store the `GraphProto.name`. I will try to exand on my original comment.
> By using the name as the graph identifier, you will only be able to have one instance, since node names should be unique. I think it would be better to use an attribute, e.g. graph_name.

### Store graph name in NodeProto.op_type
Using `NodeProto.op_type` would look like this:
```python
onnx.helper.make_node(
    op_type = model.graph.name,
    domain = "",
    name = "any_valid_c_identifier_for_diagnostic_purposes_only",
    inputs = ["a", "b", "c"],
    outputs = ["x", "y", "z"],
)
```
That conflates names from the `Graph` namespace (`GraphProto.name`) and the `Operator` namespace (`NodeProto.op_type`). cref https://github.com/onnx/onnx/blob/master/docs/IR.md#names-within-a-graph

Users seeing the op_type in error messages, performance logs, etc. or just trying to learn onnx will probably be confused that they can't find the operator in the Operators.md list.

Additionally, as you have pointed out, graphs cannot use names from the default operator set, and new releases of the ai.onnx operator set could add names which break existing models and graphs.

### Store the graph name in an attribute
I am suggesting an alternative: I think a `Graph` operator is a much clearer mechanism:
```python
onnx.helper.make_node(
    op_type = "Graph",
    domain = "",
    name = "any_valid_c_identifier_for_diagnostic_purposes_only",
    inputs = ["a", "b", "c"],
    outputs = ["x", "y", "z"],
    # Attributes:
    graph_name = model.graph.name,
)
```

The `Graph` operator can be listed in the Operators.md document, with clearly described semantics and limitations (e.g. currently only usable in training info to refer to the ModelProto.graph inference graph). It would be a special operator requiring runtime support similarly to how the `Gradient` operator requires runtime support.

Reply:
@linkerzhang, @gramalingam , adding a Graph operator sounds good to me. We just need to carefully define which names can be referenced by `graph_name`. Maybe in its first version, the only allowed name is `ModelProto.graph.name` but this syntax is extensible to call a Graph in a graph pool stored somewhere.

Reply:
I agree with @TMVector 's point about (not conflating) the two namespaces. Using a "Graph" operator is possible solution. However, we do also have the alternative mechanism ("functions") for representing a "callable subgraph". Do we want these two mechanisms? (I know at some point the training proposal using "functions" and then it was eliminated at some stage, and I don't fully know the history/reasoning behind that decision. I don't mean to resurrect a settled issue, but was wondering.)

Reply:
There are two reasons.
1. `Function` is an alternative of `Operator`, so it contains attributes which are not used when representing a graph.
2. Previous `TrainingInfoProto` contains initializers and `Function`. We can use `Graph` to cover both. The semantic of invoking a `Graph` is well-defined; in contrast, we need to further define how the inputs, outputs, and initializers are referenced by `Function`.